### PR TITLE
Submission comments

### PR DESCRIPTION
**Pull request by Tonkpils**
Please review #2713 before merging this as it is based on that. 

This will allow us to comment when submitting an iteration. 

Fixes https://github.com/exercism/exercism.io/issues/2588 and will allow us to fulfill https://github.com/exercism/cli/issues/225

### CODE REVIEW

**Comment by kytrinyx** [comment]
Oh, yay! Also thanks for doing the puma update. It was on my list to dig into what was going on there.

I'll review this today.

**Comment by Tonkpils** [comment]
Cool, I have a couple of REVIEW comments in there. Looking forward to your comments :dancer:

**Comment by kytrinyx** [comment]
Yesterday I got super sucked into adding some automated feedback to the Go track... and suddenly it was midnight. I've got it on my list for today!

File: lib/exercism/use_cases/attempt.rb [hunk 1]
```diff
@@ -12,12 +12,19 @@ def initialize(user, *stuff)
...
   def valid?
     !!slug && Xapi.exists?(track, slug)
   end
 
   def save
     user.submissions << submission
+    # REVIEW: is there any reason we are saving the user here?
+    # adding a submission as above saves the submission already.
+    #
     user.save
```
Comment (kytrinyx) [inline lib/exercism/use_cases/attempt.rb]: No, I think this is just leftover stuff from a previous change.

File: test/exercism/use_cases/attempt_test.rb [hunk 2]
```diff
@@ -13,6 +13,25 @@ def python_two
...
   end
 
+  def test_add_comment_to_submission
+    Xapi.stub(:exists?, true) do
+      attempt = Attempt.new(user, Iteration.new(python_two))
+      # REVIEW: we have to call save here because
+      # creating a new attempt creates an invalid submission
+      # once we assign it to the user on save, then we can create a comment.
+      # I'm wondering if it would be better to add a second argument to attempt
+      # that will add a comment to the submission if present on save.
```
Comment (kytrinyx) [inline test/exercism/use_cases/attempt_test.rb]: Yeah, I think that's a good suggestion. All of this code is kind of awful :/ It's still full of old experiments. I feel like I need to apologize every time someone sees it.

**Comment by kytrinyx** [comment]
I like where this is going. :heart:

**Comment by Tonkpils** [comment]
@kytrinyx I made the changes. Let me know what you think.

**Comment by kytrinyx** [comment]
Yeah, I like this better. Good call. Thank you!

**Pull request merged**